lib/TPP/GPU/GpuVulkanAbi.cpp

```diff
@@ -0,0 +1,244 @@
+#include "mlir/Dialect/SPIRV/Transforms/SPIRVConversion.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+#include <numeric>
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+// Returns Vulkan compatible wrapper type for a given input type.
+Type getVulkanTypeWrapper(Type type, const SPIRVTypeConverter &typeConverter,
+                          RewriterBase &rewriter) {
+  // Buffers are already Vulkan compatible.
+  if (type.isa<ShapedType>())
+    return type;
+
+  // Index has to be converted to a fixed-size integer.
+  if (type.isa<IndexType>()) {
+    auto spirvIndex = typeConverter.getIndexType().cast<spirv::SPIRVType>();
+    type = rewriter.getIntegerType(*(spirvIndex.getSizeInBytes()) * 8);
+  }
+
+  // Wrap the basic type in a buffer.
+  auto typeWrapper = MemRefType::get({1}, type);
+  return typeWrapper;
+};
+
+// Flatten memref type into 1D shape.
+Type FlattenMemrefType(MemRefType memrefType) {
+  if (memrefType.getRank() <= 1)
+    return memrefType;
+
+  // If the shape is not static, replace memref with a fully dynamic 1D shape.
+  // Otherwise, compute new flat size.
+  int64_t flatSize = ShapedType::kDynamic;
+  if (memrefType.hasStaticShape()) {
+    auto shape = memrefType.getShape();
+    flatSize = std::accumulate(shape.begin(), shape.end(), 1,
+                               std::multiplies<int64_t>());
+  }
+  auto flatMemref = MemRefType::get({flatSize}, memrefType.getElementType());
+
+  return flatMemref;
+}
+
+// Rewrites GPU function to have Vulkan compatible signature.
+// Replaces the original kernel with an adapted empty kernel.
+static void adaptGPUFuncToVulkanABI(gpu::GPUFuncOp &gpuFuncOp,
+                                    const SPIRVTypeConverter &typeConverter,
+                                    RewriterBase &rewriter) {
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(gpuFuncOp.getOperation());
+
+  // Create a new empty GPU kernel.
+  // The original logic is already captured by earlier conversion from GPU to
+  // SPIRV modules and an empty body ensures that kernel inputs can be easily
+  // modified.
+  auto funcOp = rewriter.cloneWithoutRegions(gpuFuncOp);
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(&funcOp.getRegion().emplaceBlock());
+    rewriter.create<gpu::ReturnOp>(gpuFuncOp.getLoc());
+  }
+
+  // Buffers are Vulkan compatible but need to be min 1D and max 3D shaped.
+  // Scalars cannot be passed directly to a Vulkan kernels.
+  // Wrap them into Vulkan compatible buffers.
+  auto funcType = funcOp.getFunctionType();
+  SmallVector<Type> inputs;
+  for (Type input : funcType.getInputs()) {
+    if (auto memrefType = input.dyn_cast<MemRefType>())
+      inputs.push_back(FlattenMemrefType(memrefType));
```

Comment:
This seems to be used only in conjunction with `getVulkanTypeWrapper`, and could simply be called from `getVulkanTypeWrapper` to simplify this loop.